pyserde unable to serialize a UUID inside of a dict[str, Any] in an object

Hi, the following serialization works in pyserde 0.14.1, but breaks in pyserde 0.14.2 (and later, up to 0.22.2)

```python
from uuid import UUID

@serde
@dataclass
class A:
    tags: dict[str, Any]

a = A(tags={'id': UUID('05b69d9a-3783-46ea-a811-35fa4b72ddac')})

to_dict(a)  # Works on 14.1, breaks on 14.2 (up to v0.22.2)
```

Here is the traceback:
```python
---------------------------------------------------------------------------
SerdeError                                Traceback (most recent call last)
Cell In[35], line 1
----> 1 to_dict(a)

File ~/Library/Caches/pypoetry/virtualenvs/test-env/lib/python3.10/site-packages/serde/se.py:461, in to_dict(o, c, reuse_instances, convert_sets)
    436 def to_dict(
    437     o: Any,
    438     c: Optional[Type[Any]] = None,
    439     reuse_instances: Optional[bool] = None,
    440     convert_sets: Optional[bool] = None,
    441 ) -> Dict[Any, Any]:
    442     """
    443     Serialize object into dictionary.
    444 
   (...)
    459     [{'i': 10, 's': 'foo', 'f': 100.0, 'b': True}, {'i': 20, 's': 'foo', 'f': 100.0, 'b': True}]
    460     """
--> 461     return to_obj(  # type: ignore
    462         o, named=True, c=c, reuse_instances=reuse_instances, convert_sets=convert_sets
    463     )

File ~/Library/Caches/pypoetry/virtualenvs/test-env/lib/python3.10/site-packages/serde/se.py:389, in to_obj(o, named, reuse_instances, convert_sets, c)
    386     return o
    388 except Exception as e:
--> 389     raise SerdeError(e) from None

SerdeError: maximum recursion depth exceeded while calling a Python object
```

If I add this as a test to `pyserde`, I get the following traceback
```python
DEBUG    serde:se.py:285 Foo: __serde__ ==================================================
                       Foo
==================================================

--------------------------------------------------
          Functions generated by pyserde
--------------------------------------------------

def to_iter(obj, reuse_instances=None, convert_sets=None, skip_none=False):
  if reuse_instances is None:
    reuse_instances = True
  if convert_sets is None:
    convert_sets = False
  if not is_dataclass(obj):
    return copy.deepcopy(obj)

  return (
  
  {k: to_obj(v, True, False, False, c=typing.Any) for k, v in obj.a.items()},
  )

def to_dict(obj, reuse_instances = None, convert_sets = None, skip_none = False):
  if reuse_instances is None:
    reuse_instances = True
  if convert_sets is None:
    convert_sets = False
  if not is_dataclass(obj):
    return copy.deepcopy(obj)

  res = {}
  subres = {k: to_obj(v, True, False, False, c=typing.Any) for k, v in obj.a.items()}
  if skip_none:
    if subres is not None:
      res["a"] = subres
  else:
    res["a"] = subres
    

  return res

--------------------------------------------------
           Function references in scope
--------------------------------------------------
to_iter: <function to_iter at 0x12ebced40>
to_dict: <function to_dict at 0x12ebcedd0>

DEBUG    serde:de.py:343 Foo: __serde__ ==================================================
                       Foo
==================================================

--------------------------------------------------
          Functions generated by pyserde
--------------------------------------------------

def to_iter(obj, reuse_instances=None, convert_sets=None, skip_none=False):
  if reuse_instances is None:
    reuse_instances = True
  if convert_sets is None:
    convert_sets = False
  if not is_dataclass(obj):
    return copy.deepcopy(obj)

  return (
  
  {k: to_obj(v, True, False, False, c=typing.Any) for k, v in obj.a.items()},
  )

def to_dict(obj, reuse_instances = None, convert_sets = None, skip_none = False):
  if reuse_instances is None:
    reuse_instances = True
  if convert_sets is None:
    convert_sets = False
  if not is_dataclass(obj):
    return copy.deepcopy(obj)

  res = {}
  subres = {k: to_obj(v, True, False, False, c=typing.Any) for k, v in obj.a.items()}
  if skip_none:
    if subres is not None:
      res["a"] = subres
  else:
    res["a"] = subres
    

  return res

def literal_de_(cls=cls, maybe_generic=None, maybe_generic_type_vars=None, data=None,
             variable_type_args=None, reuse_instances = True):
  if data in ():
    return data
  raise SerdeError("Can not deserialize " + repr(data) + " as Literal[].")
  

def from_iter(cls=cls, maybe_generic=None, maybe_generic_type_vars=None, data=None,
             variable_type_args=None, reuse_instances=None):
  if reuse_instances is None:
    reuse_instances = True

  maybe_generic_type_vars = maybe_generic_type_vars or None

  
  __a = {k: v for k, v in data[0].items()}
  

  try:
    return cls(
      
      __a,
      
    )
  except BeartypeCallHintParamViolation as e:
    raise SerdeError(e)
  except Exception as e:
    raise UserError(e)

def from_dict(cls=cls, maybe_generic=None, maybe_generic_type_vars=None, data=None,
             variable_type_args=None, reuse_instances=None):
  if reuse_instances is None:
    reuse_instances = True

  

  maybe_generic_type_vars = maybe_generic_type_vars or None

  
  __a = {k: v for k, v in data["a"].items()}
  

  try:
    return cls(
    
    
    __a,
    
    
    )
  except BeartypeCallHintParamViolation as e:
    raise SerdeError(e)
  except Exception as e:
    raise UserError(e)

--------------------------------------------------
           Function references in scope
--------------------------------------------------
to_iter: <function to_iter at 0x12ebced40>
to_dict: <function to_dict at 0x12ebcedd0>
literal_de_: <function literal_de_ at 0x12ebcf010>
from_iter: <function from_iter at 0x12ebcf0a0>
from_dict: <function from_dict at 0x12ebcf130>

DEBUG    serde:core.py:136 Generating a wrapper class code for WrapperAny
DEBUG    serde:se.py:285 WrapperAny: __serde__ ==================================================
                    WrapperAny
==================================================

--------------------------------------------------
          Functions generated by pyserde
--------------------------------------------------

def to_iter(obj, reuse_instances=None, convert_sets=None, skip_none=False):
  if reuse_instances is None:
    reuse_instances = True
  if convert_sets is None:
    convert_sets = False
  if not is_dataclass(obj):
    return copy.deepcopy(obj)

  return (
  
  to_obj(obj.v, True, False, False, c=typing.Any),
  )

def to_dict(obj, reuse_instances = None, convert_sets = None, skip_none = False):
  if reuse_instances is None:
    reuse_instances = True
  if convert_sets is None:
    convert_sets = False
  if not is_dataclass(obj):
    return copy.deepcopy(obj)

  res = {}
  subres = to_obj(obj.v, True, False, False, c=typing.Any)
  if skip_none:
    if subres is not None:
      res["v"] = subres
  else:
    res["v"] = subres
    

  return res

--------------------------------------------------
           Function references in scope
--------------------------------------------------
to_iter: <function to_iter at 0x12ebcfb50>
to_dict: <function to_dict at 0x12ebcfbe0>

DEBUG    serde:de.py:343 WrapperAny: __serde__ ==================================================
                    WrapperAny
==================================================

--------------------------------------------------
          Functions generated by pyserde
--------------------------------------------------

def to_iter(obj, reuse_instances=None, convert_sets=None, skip_none=False):
  if reuse_instances is None:
    reuse_instances = True
  if convert_sets is None:
    convert_sets = False
  if not is_dataclass(obj):
    return copy.deepcopy(obj)

  return (
  
  to_obj(obj.v, True, False, False, c=typing.Any),
  )

def to_dict(obj, reuse_instances = None, convert_sets = None, skip_none = False):
  if reuse_instances is None:
    reuse_instances = True
  if convert_sets is None:
    convert_sets = False
  if not is_dataclass(obj):
    return copy.deepcopy(obj)

  res = {}
  subres = to_obj(obj.v, True, False, False, c=typing.Any)
  if skip_none:
    if subres is not None:
      res["v"] = subres
  else:
    res["v"] = subres
    

  return res

def literal_de_(cls=cls, maybe_generic=None, maybe_generic_type_vars=None, data=None,
             variable_type_args=None, reuse_instances = True):
  if data in ():
    return data
  raise SerdeError("Can not deserialize " + repr(data) + " as Literal[].")
  

def from_iter(cls=cls, maybe_generic=None, maybe_generic_type_vars=None, data=None,
             variable_type_args=None, reuse_instances=None):
  if reuse_instances is None:
    reuse_instances = True

  maybe_generic_type_vars = maybe_generic_type_vars or None

  
  __v = data[0]
  

  try:
    return cls(
      
      __v,
      
    )
  except BeartypeCallHintParamViolation as e:
    raise SerdeError(e)
  except Exception as e:
    raise UserError(e)

def from_dict(cls=cls, maybe_generic=None, maybe_generic_type_vars=None, data=None,
             variable_type_args=None, reuse_instances=None):
  if reuse_instances is None:
    reuse_instances = True

  

  maybe_generic_type_vars = maybe_generic_type_vars or None

  
  __v = data["v"]
  

  try:
    return cls(
    
    
    __v,
    
    
    )
  except BeartypeCallHintParamViolation as e:
    raise SerdeError(e)
  except Exception as e:
    raise UserError(e)

--------------------------------------------------
           Function references in scope
--------------------------------------------------
to_iter: <function to_iter at 0x12ebcfb50>
to_dict: <function to_dict at 0x12ebcfbe0>
literal_de_: <function literal_de_ at 0x12ebcfeb0>
from_iter: <function from_iter at 0x12ebcff40>
from_dict: <function from_dict at 0x12ec14040>

DEBUG    serde:core.py:143 (de)serializing code for WrapperAny was generated
============================================================================================ short test summary info =============================================================================================
FAILED tests/test_basics.py::test_uuid_in_dict_in_obj - serde.compat.SerdeError: maximum recursion depth exceeded
```

--- a/serde/se.py
+++ b/serde/se.py
@@ -381,8 +381,9 @@ def to_obj(
         elif is_bearable(o, dict):
             return {k: thisfunc(v) for k, v in o.items()}
         elif is_str_serializable_instance(o) or is_datetime_instance(o):
+            se_cls = o.__class__ if not c or c is Any else c
             return CACHE.serialize(
-                c or o.__class__,
+                se_cls,
                 o,
                 reuse_instances=reuse_instances,
                 convert_sets=convert_sets,

--- a/tests/test_basics.py
+++ b/tests/test_basics.py
@@ -662,7 +662,6 @@ def test_skip_if_default(se, de):
 
 @pytest.mark.parametrize("se,de", format_msgpack)
 def test_inheritance(se, de):
-
     @serde.serde
     class Base:
         a: int
@@ -1113,3 +1112,23 @@ def test_init_var_with_field_attribute() -> None:
         a: int = serde.field(skip=True)
 
     assert serde.to_dict(Foo(10)) == {}
+
+
+def test_dict_str_any() -> None:
+    @serde.serde
+    class Foo:
+        a: dict[str, Any]
+
+    # Because the dict values are typed as Any, all values will be serialized as strings
+    date = datetime.datetime(2024, 12, 3, 16, 55, 20, 220662)
+    date_str = date.isoformat()
+
+    uuid_val = uuid.UUID("efb71d76-4337-4b27-a612-5b6f95b01d42")
+    uuid_str = str(uuid_val)
+
+    foo = Foo(a={"a": uuid_val, "b": date})
+    foo_se = {"a": {"a": uuid_str, "b": date_str}}
+    foo_de = Foo(a={"a": uuid_str, "b": date_str})
+
+    assert serde.to_dict(foo) == foo_se
+    assert serde.from_dict(Foo, foo_se) == foo_de

